core/src/commonMain/kotlin/com/powersync/attachments/AttachmentQueue.kt

```diff
@@ -0,0 +1,489 @@
+                        // The attachment is present again. Need to queue it for sync.
+                        // We might be able to optimize this in future
+                        if (existingQueueItem.hasSynced == 1) {
+                            // No remote action required, we can restore the record (avoids deletion)
+                            attachmentUpdates.add(
+                                existingQueueItem.copy(state = AttachmentState.SYNCED.ordinal),
+                            )
+                        } else {
+                            /**
+                             * The localURI should be set if the record was meant to be downloaded
+                             * and has been synced. If it's missing and hasSynced is false then
+                             * it must be an upload operation
+                             */
+                            attachmentUpdates.add(
+                                existingQueueItem.copy(
+                                    state =
+                                        if (existingQueueItem.localUri == null) {
+                                            AttachmentState.QUEUED_DOWNLOAD.ordinal
+                                        } else {
+                                            AttachmentState.QUEUED_UPLOAD.ordinal
+                                        },
+                                ),
+                            )
+                        }
+                    }
+                }
+
+                /**
+                 * Archive any items not specified in the watched items except for items pending delete.
+                 */
+                currentAttachments
+                    .filter {
+                        it.state != AttachmentState.QUEUED_DELETE.ordinal &&
+                            null == items.find { update -> update.id == it.id }
+                    }.forEach {
+                        attachmentUpdates.add(it.copy(state = AttachmentState.ARCHIVED.ordinal))
+                    }
+
+                attachmentsContext.saveAttachments(attachmentUpdates)
+            }
+        }
+
+    /**
+     * A function which creates a new attachment locally. This new attachment is queued for upload
+     * after creation.
+     *
+     * The filename is resolved using [resolveNewAttachmentFilename].
+     *
+     * A [updateHook] is provided which should be used when assigning relationships to the newly
+     * created attachment. This hook is executed in the same writeTransaction which creates the
+     * attachment record.
+     *
+     * This method can be overriden for custom behaviour.
+     */
+    @Throws(PowerSyncException::class, CancellationException::class)
+    public open suspend fun saveFile(
+        data: Flow<ByteArray>,
+        mediaType: String,
+        fileExtension: String? = null,
+        metaData: String? = null,
+        updateHook: (context: ConnectionContext, attachment: Attachment) -> Unit,
```

Comment:
I wonder if this should be suspending just in case

Reply:
This hook currently is executed inside the write transaction which is unfortunately not a suspending context.